va/caaQuorum.go

```diff
@@ -0,0 +1,246 @@
+	"net"
+	"net/http"
+	"net/url"
+	"sort"
+	"time"
+
+	"github.com/miekg/dns"
+	"golang.org/x/net/context"
+	"golang.org/x/net/context/ctxhttp"
+
+	"github.com/letsencrypt/boulder/metrics"
+)
+
+// We have found a number of network operators which block or drop CAA
+// queries that pass through their network which leads to consistent
+// timeout failures from certain network perspectives. We have been
+// unable to find a network solution to this so we are required to
+// implement a multi-path resolution technique. This is a real hack and
+// to be honest probably not the best solution to this problem. Ideally
+// we would control our own distributed multi-path resolver but there
+// are no publicly available ones.
+//
+// This implementation talks to the Google Public DNS resolver over
+// multiple paths using HTTP proxies with geographically distributed
+// endpoints. In case the Google resolver encounters the same issues we do
+// multiple queries for the same name in parallel and we require a M of N
+// quorum of responses to return the SUCCESS return code. In order to prevent
+// the case where an attacker may be able to exploit the Google resolver in
+// some way we also require that the records returned from all requests are
+// the same (as far as I can tell the Google DNS implementation doesn't share
+// cache state between the distributed nodes so this should be safe).
+//
+// Since DNS isn't a super secure protocol and Google has recently introduced
+// a public HTTPS API for their DNS resolver so instead we use that.
+//
+// API reference:
+//   https://developers.google.com/speed/public-dns/docs/dns-over-https#api_specification
+
+var apiURI = "https://dns.google.com/resolve"
+
+type question struct {
+	Name string `json:"name"`
+	Type uint16 `json:"type"`
+}
+
+type answer struct {
+	Name string `json:"name"`
+	Type uint16 `json:"type"`
+	TTL  int    `json:"TTL"`
+	Data string `json:"data"`
+}
+
+type response struct {
+	// Ignored fields
+	//   tc
+	//   rd
+	//   ra
+	//   ad
+	//   cd
+	//   question
+	//   additional
+	//   edns_client_subnet
+	Status  int      `json:"Status"`
+	Answer  []answer `json:"Answer"`
+	Comment string   `json:"Comment"`
+}
+
+func parseAnswer(as []answer) ([]*dns.CAA, error) {
+	rrs := []*dns.CAA{}
+	// only bother parsing out CAA records
+	for _, a := range as {
+		if a.Type != 257 {
+			continue
+		}
+		rr, err := dns.NewRR(fmt.Sprintf("%s %d IN %s %s", a.Name, a.TTL, dns.TypeToString[a.Type], a.Data))
+		if err != nil {
+			return nil, err
+		}
+		if caaRR, ok := rr.(*dns.CAA); ok {
+			rrs = append(rrs, caaRR)
+		}
+	}
+	return rrs, nil
+}
+
+func createClient(timeout, keepAlive time.Duration, proxy string) (*http.Client, string, error) {
+	u, err := url.Parse(proxy)
+	if err != nil {
+		return nil, "", err
+	}
+	transport := &http.Transport{
+		Proxy: http.ProxyURL(u),
+		Dial: (&net.Dialer{
+			Timeout:   timeout,
+			KeepAlive: keepAlive,
+		}).Dial,
+		TLSHandshakeTimeout: 10 * time.Second,
```

Comment:
We currently don't set this anywhere else (i.e. in VA code) so I just jumped to using the default.